src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/EditorStateModel.cs

```diff
@@ -0,0 +1,289 @@
+            }
+            _textBufferManager?.Dispose();
+
+            _projectFileModelWatcher = null;
+            _frameEventsListener = null;
+            _textBufferStateListener = null;
+            _textBufferManager = null;
+
+            lock (_lock)
+            {
+                _currentState = EditorState.NoEditor;
+            }
+            return;
+        }
+
+        // Update Project File Logic
+
+        /// <summary>
+        /// Schedules the project file to be updated. This is called by the ProjectFileModelWatcher, which is within a Project write lock.
+        /// Because of the time-sensitive nature of that lock, instead of updating the file immediately we schedule an update as soon
+        /// as the JTF can run our code.
+        /// </summary>
+        public JoinableTask ScheduleProjectFileUpdate()
+        {
+            lock (_lock)
+            {
+                // If the current state is writing project file, we don't want to update now, as the project will not have been fully
+                // reloaded yet. We'll get called back again after the ProjectReloadManager is finished reloading the project.
+                // Additionally, if there is no editor open, or if the editor is closing, don't schedule an update
+                if (_currentState == EditorState.WritingProjectFile || _currentState == EditorState.NoEditor || _currentState == EditorState.EditorClosing) return null;
+                _currentState = EditorState.BufferUpdateScheduled;
+                return _threadingService.JoinableTaskFactory.RunAsync(UpdateProjectFileAsync);
+            }
+        }
+
+        /// <summary>
+        /// Updates the content of the project file to be the latest msbuild text.
+        /// </summary>
+        /// <returns></returns>
+        private async Task UpdateProjectFileAsync()
+        {
+            // We do this on the UI thread to ensure that between setting the state to a variant of ProjectFileChanging and setting the window
+            // frame to readonly, the user can't do any input (and potentially cause a conflict).
+            await _threadingService.SwitchToUIThread();
+            lock (_lock)
+            {
+                // Only update the project file if we have scheduled an update. Any other state is either already updating the buffer,
+                // or in the processes of closing the buffer.
+                if (_currentState != EditorState.BufferUpdateScheduled)
+                {
+                    return;
+                }
+            }
+
+            // Set set the buffer to be unmodifiable to prevent any changes while we update the project
+            await _textBufferManager.SetReadOnlyAsync(true).ConfigureAwait(true);
+
+            // Set the buffer to default state and set the current state to editor clean
+            await _textBufferManager.ResetBufferAsync().ConfigureAwait(true);
```

Comment:
Yes. `ResetBufferAsync` puts the buffer in sync with the project file, correctly handling if the buffer has unsaved changes.